dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp

```
@@ -481,33 +482,101 @@ SegmentPtr DeltaMergeStore::segmentMergeDelta(
+        // Note: applyIngestData itself writes the wbs, and we don't need to pass a wbs by ourselves.
+        auto apply_result = segment->applyIngestData(segment_lock, dm_context, data_file, ingest_info);
 
-        RUNTIME_CHECK(compare(segment->getRowKeyRange().getEnd(), new_segment->getRowKeyRange().getEnd()) == 0, segment->info(), new_segment->info());
-        RUNTIME_CHECK(segment->segmentId() == new_segment->segmentId(), segment->info(), new_segment->info());
-
-        segment->abandon(dm_context);
-        segments[segment->getRowKeyRange().getEnd()] = new_segment;
-        id_to_segment[segment->segmentId()] = new_segment;
+        if (auto * const result = std::get_if<Segment::IngestDataResultSegmentReplaced>(&apply_result); result)
+        {
+            new_segment = result->segment;
+
+            RUNTIME_CHECK(
+                compare(segment->getRowKeyRange().getEnd(), new_segment->getRowKeyRange().getEnd()) == 0,
+                segment->info(),
+                new_segment->info());
+            RUNTIME_CHECK(
+                segment->segmentId() == new_segment->segmentId(),
+                segment->info(),
+                new_segment->info());
+
+            segment->abandon(dm_context);
+            segments[segment->getRowKeyRange().getEnd()] = new_segment;
+            id_to_segment[segment->segmentId()] = new_segment;
+
+            LOG_INFO(
+                log,
+                "IngestData - Finish, new segment is created, old_segment={} new_segment={}",
+                segment->info(),
+                new_segment->info());
+        }
+        else if (std::holds_alternative<Segment::IngestDataResultError>(apply_result))
+        {
+            // This should not happen, because we have verified segment is not abandoned.
+            LOG_DEBUG(log, "IngestData - Give up segmentIngestData because applyIngestData failed (usually this should not happen), segment={}", segment->simpleInfo());
+            return {};
```

We should throw an exception or at least use `LOG_ERROR` here to indicate something went wrong. Or it may run into an inf retrying loop without any clear error message.